calico/base.py

```diff
@@ -328,3 +334,82 @@ def run(self, tests=None, quiet=False, g_timeout=None):
 
         report["points"] = earned_points
         return report
+
+
+class Clioc:
+    """A class that is able to generate test specifications from a reference program's runs."""
+
+    def __init__(self, argv):
+        """Initialize this test specification generator.
+
+        :sig: (List[str]) -> None
+        :param argv: List of command line arguments to run the reference program.
+        """
+        self.argv = argv
+        """"""
+        self.__current_test_case = None
+
+    def generate_test_spec(self):
+        """
+        Run the reference program and return the generated test specification.
+
+        :sig: () -> str
+        :return: The specification that is generated from reference program's run.
+        """
+        test_suite = self.__create_test_suite()
+        return str(test_suite)
+
+    def __create_test_suite(self):
+        calico = Calico()
+        case_number = count(1)
+        while True:
+            os.system("clear")
+            calico.add_case(self.__create_test_case(next(case_number)))
+            if input("Do you want to continue? [Y/n] ").lower() not in ("y", ""):
+                print("Abort.")
+                break
+        return calico
+
+    def __create_test_case(self, case_num):
+        case_name = f"case_{case_num}"
+        print(f"Running for {case_name}...")
+        self.__current_test_case = TestCase(case_name, " ".join(self.argv), None)
+        exit_code = self.__spawn()
+        self.__current_test_case.add_action(Action(ActionType.EXPECT, "_EOF_"))
+        print(f"{case_name} ended")
+        points = int(input("Assign points for this run: "))
+        self.__current_test_case.points = points
+        self.__current_test_case.exits = exit_code
+        return self.__current_test_case
+
+    def __spawn(self):
+        pid, master_fd = pty.fork()
+        if pid == pty.CHILD:
+            os.execlp(self.argv[0], *self.argv)
+        try:
+            mode = tty.tcgetattr(pty.STDIN_FILENO)
+            tty.setraw(master_fd, termios.TCSANOW)
+        except tty.error:
+            restore = False
+        else:
+            restore = True
+
+        try:
+            pty._copy(master_fd, self.__read_write_handler, self.__read_write_handler)
+        except OSError:
+            if restore:
+                tty.tcsetattr(pty.STDIN_FILENO, tty.TCSAFLUSH, mode)
+
+        os.close(master_fd)
+        return os.waitpid(pid, 0)[1] >> 8
```

Comment:
This function is almost exact copy of [`pty.spawn`](https://github.com/python/cpython/blob/master/Lib/pty.py#L151) with a difference on one line: `tty.setraw(master_fd, termios.TCSANOW)`. It puts the newly opened pseudo-terminal into raw mode and I had no idea why this solves the problem I mentioned [here](https://stackoverflow.com/q/62288531/9608759). [`tty.setraw`](https://github.com/python/cpython/blob/master/Lib/tty.py#L18) sets some flags for the given file descriptor and only few of them are related with our problem. These are the lines that I found necessary for us after a bit of debugging:

```python
mode[OFLAG] = mode[OFLAG] & ~(OPOST)
mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON | IEXTEN | ISIG) # and only `~ECHO` is enough here actually
```
> `OPOST`: Enable implementation-defined output processing.

This, for example when enabled puts a carriage return, `\r`, before newline character`\n`. Since it's negated in the above code, it treats newline as a newline and I had to add it manually afterwards (413th line in this file). So I think we don't need this line. (`OPOST` will be set, we don't need to add `\r` manually)

> `ECHO`: Echo input characters.

This will make the pseudo-terminal device to echo it's stdin when enabled. We already capture it through our stdin, so we should disable echoing of the same characters. This line is needed. 

The other flags, (`ICANON`, `IEXTEN`, `ISIG`), have no effect on the result as far as I have observed.


I also realized that `try ... except ...` thing for restoring has no effect because we don't change anything for the file descriptor, `pty.STDIN_FILENO`. We can also delete these lines.

The purpose of this review was to mention [`pty.spawn`](https://github.com/python/cpython/blob/master/Lib/pty.py#L151) to prove the correctness of my function with a suggestion to delete `try ... except ...` statement but it ended up explaining all the shady parts. I'll update the code with the changes. Hopefully, it will become much simpler.